sensor_msgs/include/sensor_msgs/point_cloud_msg_wrapper.hpp

```diff
@@ -230,17 +231,26 @@ class PointCloudMsgWrapper
   COMPILE_IF_MUTABLE(CloudMsgT, void) push_back(const PointT & point)
   {
     PointT point_copy{point};
-    extend_data_by(m_cloud_ref.point_step);
-    m_cloud_ref.row_step += m_cloud_ref.point_step;
+    extend_data_by(sizeof(PointT));
+    m_cloud_ref.row_step += sizeof(PointT);
     m_cloud_ref.width++;
     this->operator[](m_cloud_ref.width - 1U) = point_copy;
   }
 
+  /// @brief      Push a new point into the message.
+  COMPILE_IF_MUTABLE(CloudMsgT, void) push_back_2(const PointT & point)
```

Comment:
Micro-benchmarking is hard. If you haven't seen [this talk on cppcon](https://www.youtube.com/watch?v=nXaxk27zwlk) I highly recommend it. Basically, it is very hard to measure exactly what we want in such benchmarks. That is why I generally don't trust micro-benchmarking unless it is written as part of Google Benchmark suite (it seems to be available in ros2 rolling if you'd like to give it a go) and a all the corner cases have been taken care of.

And to illustrate what I'm talking about, In this particular situation, when I build the code in release mode: 
```
 colcon build --symlink-install --merge-install --packages-up-to sensor_msgs --cmake-args -DCMAKE_BUILD_TYPE=Release
 ```
 I get the following output from your test: 
 ```
push_back took: 1671.54 ms.
push_back_2 took: 1713.92 ms.
```
That being said, if I just change the order of the tests (run `push_back_2` before `push_back`) then the situation is reversed:
```
push_back_2 took: 1640.54 ms.
push_back took: 1744.65 ms.
```
So the moral of this is the following: in my opinion, unless we want to invest in a proper benchmarking done through google-benchmark (which we might actually want to do) we should prefer readability over micro-optimization. What do you think about it?

Reply:
I don't really know the specifics of how GTest is running these tests in order or how parallel so I cannot comment on why it produces such different time results.

I generally put both implementations side by side and check the timings with high_resolution_clock like I did in the test and generally have consistent results but in this instance I didn't bother to create a node that makes use of this class and measure times there and just wanted to get it all done in the same package.

> Micro-benchmarking is hard. If you haven't seen [this talk on cppcon](https://www.youtube.com/watch?v=nXaxk27zwlk) I highly recommend it.

Will definitely watch it!

> So the moral of this is the following: in my opinion, unless we want to invest in a proper benchmarking done through google-benchmark (which we might actually want to do) we should prefer readability over micro-optimization. What do you think about it?

Well the whole point of this library is to not have copy overhead of pcl_conversions library thus having everything as fast as possible. So if something makes it faster it should be in there whether it is more or less readable in the library.

- The push_back you implemented does resize the data vector by +1 and assigns last value to a copy of the point's memory address.
- What I propose reinterprets the given point reference as byte vector and inserts it to the end of data vector.

I don't have the right tooling to check ASM representations of the both codes (also they would probably be distorted with the optimizations in based on where they'd be used) so cannot say which is better without doing benchmarks. But I definitely understand that you don't want to sacrifice readability before seeing consistent speed gains, it's alright.

I am just getting familiar with gtest and google benchmark, not quite adept on using either so If you want, you can benchmark them.